### PR TITLE
Add AMJ66 keyboard

### PR DESCRIPTION
**Pull request by noroadsleft**
## Description

Add the AMJ66 PCB under the KBDfans directory.

## Notes

Based on work by @FSund and @fauxpark.

### CODE REVIEW

**Comment by noroadsleft** [comment]
## To Do

- [ ] Solve the backlighting weirdness
- [ ] Add more layout macros (ANSI, ISO)
- [ ] Configurator support
- [ ] Give FSund an actual usable keymap
- [ ] Proof the readme files

## Questions

- Should this actually go in the KBDfans directory?

File: keyboards/kbdfans/amj66/config.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+/* Keyboard Matrix Assignments */
+#define MATRIX_ROW_PINS { F7, F6, F5, F4, F1 } // from judy2k amj66 firmware
+#define MATRIX_COL_PINS { F0, B3, B2, B1, B0, B7, D0, D1, D2, D3, D5, D6, D7, B4, B5, B6 }
+#define UNUSED_PINS
+
+/* COL2ROW, ROW2COL*/
+#define DIODE_DIRECTION COL2ROW
+
+// #define BACKLIGHT_PIN D4
```
Comment (fauxpark) [inline keyboards/kbdfans/amj66/config.h]:
```suggestion
#define BACKLIGHT_PIN D4
```

File: keyboards/kbdfans/amj66/config.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+// #define BACKLIGHT_LEVELS 3
```
Comment (fauxpark) [inline keyboards/kbdfans/amj66/config.h]:
```suggestion
#define BACKLIGHT_LEVELS 3
```

File: keyboards/kbdfans/amj66/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    [0] = LAYOUT_all(
+        KC_ESC,           KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_BSPC,     KC_INS,
+        KC_TAB,           KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS,              KC_DEL,
+        KC_CAPS,          KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,          KC_ENT,
+        KC_LSFT, KC_NUBS, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_RSFT, MO(1),   KC_UP,
+        KC_LCTL, KC_LGUI, KC_LALT,                   KC_SPC,  KC_SPC,                    KC_RALT, KC_RGUI, KC_RCTL,          KC_LEFT, KC_DOWN, KC_RGHT
+    ),
+
+    [0] = LAYOUT_all(
```
Comment (fauxpark) [inline keyboards/kbdfans/amj66/keymaps/default/keymap.c]:
```suggestion
    [1] = LAYOUT_all(
```

File: keyboards/kbdfans/amj66/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        KC_GRV,           KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,  KC_BSPC, KC_BSPC,     KC_INS,
+        _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, KC_PSRC, KC_SLCK, KC_PAUS,              KC_DEL,
```
Comment (fauxpark) [inline keyboards/kbdfans/amj66/keymaps/default/keymap.c]:
```suggestion
        _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, KC_PSCR, KC_SLCK, KC_PAUS,              KC_DEL,
```

File: keyboards/kbdfans/amj66/config.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+// #define BACKLIGHT_BREATHING
```
Comment (fauxpark) [inline keyboards/kbdfans/amj66/config.h]:
```suggestion
#define BACKLIGHT_BREATHING
```

File: keyboards/kbdfans/amj66/config.h [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+// #define BREATHING_PERIOD 6
```
Comment (fauxpark) [inline keyboards/kbdfans/amj66/config.h]:
```suggestion
#define BREATHING_PERIOD 6
```

File: keyboards/kbdfans/amj66/rules.mk [hunk 7]
```diff
@@ -0,0 +1,68 @@
+# MCU name
+#MCU = at90usb1286
```
Comment (fauxpark) [inline keyboards/kbdfans/amj66/rules.mk]:
```suggestion
```

File: keyboards/kbdfans/amj66/rules.mk [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+
+# Boot Section Size in *bytes*
+#   Teensy halfKay   512
+#   Teensy++ halfKay 1024
+#   Atmel DFU loader 4096
+#   LUFA bootloader  4096
+#   USBaspLoader     2048
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
```
Comment (fauxpark) [inline keyboards/kbdfans/amj66/rules.mk]: This should be:

```make
# Bootloader selection
#   Teensy       halfkay
#   Pro Micro    caterina
#   Atmel DFU    atmel-dfu
#   LUFA DFU     lufa-dfu
#   QMK DFU      qmk-dfu
#   atmega32a    bootloadHID
BOOTLOADER = atmel-dfu
```

File: keyboards/kbdfans/amj66/amj66.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+ * │10   │11 │12 │13 │14 │15 │16 │17 │18 │19 │1A │1B │1C │1D   │ │1E │      │2C   │
+ * ├─────┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴┬──┴─────┤ └───┘   ┌──┴┐    │
+ * │20    │21 │22 │23 │24 │25 │26 │27 │28 │29 │2A │2B │2C      │         │1D │    │
+ * ├────┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴─┬─┴──┬───┬─┴─┐       └───┴────┘
+ * │30  │31 │32 │33 │34 │35 │36 │37 │38 │39 │3A │3B │3C  │3D │3E │
+ * ├────┴┬──┴─┬─┴──┬┴───┴───┴─┬─┴───┴───┴┬──┴─┬─┴──┬┴────┼───┼───┼───┐
+ * │40   │41  │42  │43        │44        │45  │46  │47   │48 │49 │4A │
+ * └─────┴────┴────┴──────────┴──────────┴────┴────┴─────┴───┴───┴───┘
+ */
+#define LAYOUT_all( \
```
Comment (fauxpark) [inline keyboards/kbdfans/amj66/amj66.h]: Probably could be split into ISO and ANSI layouts

**Comment by fauxpark** [comment]
I tried compiling this with `BACKLIGHT_PIN D4`, and it produced:

```
quantum/quantum.c:1063:9: note: #pragma message: Using hardware timer 1 with software PWM
 #pragma message "Using hardware timer 1 with software PWM"
         ^~~~~~~
```

👍

File: keyboards/kbdfans/amj66/keymaps/default/keymap.c (hunk 3 above)
Comment (noroadsleft) [reply keyboards/kbdfans/amj66/keymaps/default/keymap.c]: *derp*

**Comment by noroadsleft** [comment]
I'll hopefully be picking this back up "tomorrow"; if not, then Wednesday for sure. It's now nearly 4:30 AM for me, and I have a pretty full day ahead.

**Comment by FSund** [comment]
This will fix #5930, and the LED issues I'm experiencing are probably related to #5425.

**Comment by mechmerlin** [comment]
Just throwing in more info here: https://github.com/AMJKeyboard/AMJ66

**Comment by noroadsleft** [comment]
@mechmerlin Yeah, I was leaning toward not in the KBDfans directory, but @FSund already had his files there, so for simplicity I left them there.

If we reach a consensus to have the board in its own directory or in an AMJ directory, I'll do that last once all the other parts are done.

**Comment by FSund** [comment]
I didn't realize there were so many AMJ keyboards. In that case I think a common AMJ folder would probably be a good solution. But there should probably be a note/warning in the kbd66 and amj66 readme's about the different versions.

**Comment by fauxpark** [comment]
@FSund confirmed the backlight on this board is working after setting `BACKLIGHT_ON_STATE 1` and applying #5983, so I believe that should be added here.

**Comment by noroadsleft** [comment]
Applied the `BACKLIGHT_ON_STATE` definition.

This PR will require #5983 to work fully, so I've left that file unchanged so as to prevent conflicts.

Also co-credited @fauxpark and @FSund on the last commit because they've done a lot of legwork on this PR and deserve to be credited more than I do, IMO.

**Comment by noroadsleft** [comment]
I seem to have missed or forgotten that #5983 was merged. If @fauxpark or @FSund can confirm functionality on this as is, we should be good to merge this in.

**Comment by fauxpark** [comment]
I don't have this board but it should be good to go.

**Comment by FSund** [comment]
Sorry, have been busy this summer. Will test this out as soon as I'm back with my AMJ66 board.

**Comment by FSund** [comment]
Got around to testing the AMJ66 board now.

First something strange I've never seen before. When compiling the firmware it also compiles a couple of keymaps that I can't find anywhere in the amj66 folder, namely a `mechmerlin` and a `skully` keymap.

    $ make kbdfans/amj66
    QMK Firmware 0.6.385
    WARNING:
     Some git sub-modules are out of date or modified, please consider running:
     make git-submodule
     You can ignore this warning if you are not compiling any ChibiOS keyboards,
     or if you have modified the ChibiOS libraries yourself.

    Making kbdfans/amj66 with keymap default                                                               [OK]
    Making kbdfans/amj66 with keymap default_66_ansi                                                       [OK]
    Making kbdfans/amj66 with keymap default_66_iso                                                        [OK]
    Making kbdfans/amj66 with keymap fsund_test                                                            [OK]
    Making kbdfans/amj66 with keymap mechmerlin                                                            [OK]
    Making kbdfans/amj66 with keymap skully                                                                [OK]

I only have `default` and `fsund_test` in the keymaps folder, and I just did a fresh clone into a new folder to confirm this.

Further, `kbdfans_amj66_default.hex` and `kbdfans_amj66_default_66_ansi.hex` that are created seems to work perfectly. 

But `kbdfans_amj66_default_66_iso.hex` has some issues. Left and right ALT keys are mapped to Win and vice versa, Caps Lock is mapped to Ctrl, and the two keys in the top right corner are mapped to PgUp and PgDn instead of Ins and Del as they are in the keymap.

**Comment by noroadsleft** [comment]
@FSund This is all normal.

I enabled Community Layout support:

https://github.com/qmk/qmk_firmware/blob/e277a35d16c46f42bfb2cf86047e9b8df5393d56/keyboards/kbdfans/amj66/rules.mk#L70

which makes the QMK build script check for keymaps in:

- [qmk_firmware/layouts/community/66_ansi](https://github.com/qmk/qmk_firmware/tree/e277a35d16c46f42bfb2cf86047e9b8df5393d56/layouts/community/66_ansi)
- `qmk_firmware/layouts/community/66_iso` (doesn't exist)
- [qmk_firmware/layouts/default/66_ansi](https://github.com/qmk/qmk_firmware/tree/e277a35d16c46f42bfb2cf86047e9b8df5393d56/layouts/default/66_ansi)
- [qmk_firmware/layouts/default/66_iso](https://github.com/qmk/qmk_firmware/tree/e277a35d16c46f42bfb2cf86047e9b8df5393d56/layouts/default/66_iso)


The `layouts/default/` folder is used for testing, and has reference keymaps. `layouts/community/` is for users who own multiple QMK boards that share a form factor (like owning a GH60, a DZ60, and a 1up 60 HotSwap). `layouts/community/66_ansi/` has the `mechmerlin` and `skully` keymaps, and the two folders I referenced in `layouts/default/` are the source of the `default_66_ansi` and `default_66_iso` keymaps.

The `default_66_iso` keymap does have some not-as-common assignments as you've discovered:

https://github.com/qmk/qmk_firmware/blob/e277a35d16c46f42bfb2cf86047e9b8df5393d56/layouts/default/66_iso/default_66_iso/keymap.c#L4-L10

... but that's irrelevant, as this keymaps isn't really there to be used by users anyway.

As long as your `fsund_test` keymap works as you've intended, then this codebase is working. If that's good to go, then I'll move the files to something like `keyboards/amjkeyboard/` tomorrow. (I think that's the folder I had planned on.)

**Comment by FSund** [comment]
@noroadsleft Oh, I didn't know we had community layouts. Nice! 

Just checked the LED stuff, and most of it seems to be working fine. The only issue I found was that I'm unable to turn off the breathing effect. `BL_BRTG` doesn't have any effect. Not sure if this is a separate issue with soft PWM (something not caught in [pull 5983](https://github.com/qmk/qmk_firmware/pull/5983) perhaps?), or a config issue with AMJ66. I tried calling `breathing_disable()` once in `matrix_init_kb()`, but that didn't do anything either.

**Comment by fauxpark** [comment]
That was indeed fixed in 5983, but this branch is from before I even opened that PR.

**Comment by FSund** [comment]
Ah, too much to keep track of for me. Guess we're all ready for merging then, and I'll verify that LEDs work as intended after.

**Comment by noroadsleft** [comment]
Made some more changes; mostly QMK convention stuff.

The files have been moved to a new `keyboards/amjkeyboards/` folder. (I'll move the existing AMJ keyboards in a separate PR once this is merged.)

Also rebased against the current QMK master branch.

@FSund,

This branch now includes @fauxpark's backlight fix from #5983, so you should have fully working backlight.

I've also included you as a keyboard maintainer; let me know if you'd rather not be listed.

Also let me know if you intend to submit your own pull request for your AMJ66 keymap to qmk/qmk_firmware once we're done testing. If you'd rather have it included with this PR, open a PR to [my `kb/amj66` branch](https://github.com/noroadsleft/qmk_firmware/tree/kb/amj66) with the changes needed.

**Comment by FSund** [comment]
Perfect, everything still works great. Now the LED breathing toggle works as well. I don't mind being a maintainer.

You can probably remove the fsund_test keymap.

I think I'll submit my on PR for my keymap when I'm finished fiddling with it :)

**Comment by peepeetee** [comment]
necroing this to say that I am doing what was discussed and creating an AMJ folder here:https://github.com/qmk/qmk_firmware/pull/15733

**Pull request merged**